### DISCUSSION TITLE
Iso Volume Filter needed

### DISCUSSION THREAD

**Discussion by Jimmy-KL** [pyvista/pyvista#3754] [Q&A] [answered]
@banesullivan In Paraview, there is a filter called "Iso Volume". Unlike "Threshold", it will interpolate the data within a cell for node scalar data which will give us a smooth model.
Sometimes, "Iso Volume" is more useful than "Contour", "Threshold" and "Clip" for the volumetric model. I think it's possible to do the same thing with a couple of filters, but it will be more convenient to have the "Iso Volume" filter in pyvista.
BTW, I think "Iso Volume" is clipping the data using both the upper threshold and lower threshold, right?

**Comment by banesullivan** [comment]
Sorry for the delay... this is on my backlog of things to look into as having an IsoVolume filter in PyVista would be a great addition. Unfortunately, this is an involved task so I'll have to come back to it when I have the extra bandwidth.

... stay tuned.

**Reply by kamakshidasan** [reply]
hi @banesullivan, any updates on this? :)

**Reply by banesullivan** [reply, marked as answer]
We still do not have an "isovolume" filter in PyVista as far as I'm aware but this is easy to implement as two [clip_scalar](https://docs.pyvista.org/api/core/_autosummary/pyvista.datasetfilters.clip_scalar#pyvista.DataSetFilters.clip_scalar) filters chained together:

```py
import pyvista as pv

def isovolume(mesh: pv.DataSet, vmin: float, vmax: float, scalars: str | None = None):
    """Perform an isovolume clip."""
    return mesh.clip_scalar(
        scalars=scalars, value=vmin, invert=False,
    ).clip_scalar(
        scalars=scalars, value=vmax, invert=True,
    )

# Example
mesh = pv.Wavelet()
vol = isovolume(mesh, 125, 200)
vol.plot(clim=mesh.get_data_range("RTData"))
```

![screenshot (1)](https://github.com/user-attachments/assets/c6b819b7-3dbe-46c6-b217-6824c944c0a5)


If you're feeling up for it, this would be a great and welcomed addition!

Another option is a [threshold([a, b])](https://docs.pyvista.org/api/core/_autosummary/pyvista.datasetfilters.threshold#pyvista-datasetfilters-threshold) filter with an upper and lower bound: 

```py
import pyvista as pv
mesh = pv.Wavelet()
vol = mesh.threshold([125, 200], scalars="RTData", invert=False)
vol.plot(clim=mesh.get_data_range("RTData"))
```

![screenshot (2)](https://github.com/user-attachments/assets/cc21b8c0-8fd5-42d9-b09a-08af780b248c)

**Reply by banesullivan** [reply]
Issue tracked in #7256

**Comment by kamakshidasan** [comment]
@banesullivan - I will be happy to contribute, can you point to how/where I should start?

**Reply by banesullivan** [reply]
Let's start with the discussion in https://github.com/pyvista/pyvista/issues/7256 about implementation, then if you're up for a PR to add this to the DataSetFilters class next to the existing `clip_scalar` filter would do it. Thanks for your interest in contributing!
